{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e01c7cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from scipy.stats import linregress as lrs\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "def filtering(genotypes, strains):\n",
    "    relevant_strains = [column for column in genotypes.columns if column in strains]\n",
    "    columns = list(genotypes.columns)[:3]\n",
    "    columns.extend(relevant_strains)\n",
    "    final_genotypes = genotypes[columns]\n",
    "    drop_indices = []\n",
    "    for i in range(len(final_genotypes) - 1):\n",
    "        if (final_genotypes.iloc[i, 3:]).equals(final_genotypes.iloc[i + 1, 3:]):\n",
    "            drop_indices.append(i + 1)\n",
    "    final_genotypes = final_genotypes.drop(drop_indices)\n",
    "    return final_genotypes\n",
    "\n",
    "\n",
    "def mean_data(table):\n",
    "    cols_set = {col.split('_')[0] for col in table.columns if 'BXD' in col}\n",
    "    means_table = pd.DataFrame({'data': table['data']})\n",
    "    for strain in cols_set:\n",
    "        temp_cols = [col for col in table.columns if col.split('_')[0] == strain]\n",
    "        means_table[strain] = table[temp_cols].mean(axis=1)\n",
    "    sorted_cols = list(table.columns[:1])\n",
    "    sorted_cols.extend(sorted(means_table.columns[1:], key=lambda x: int(x[3:])))\n",
    "    means_table = means_table[sorted_cols]\n",
    "    return means_table\n",
    "\n",
    "\n",
    "def association_model(genotypes_df, expression_df, alpha=0.05):\n",
    "    relevant_cols = [col for col in genotypes_df.columns if col in expression_df.columns]\n",
    "    genotypes_df = genotypes_df[genotypes_df.columns[:3].to_list() + relevant_cols]\n",
    "    strains_dict = dict()\n",
    "    for i in range(len(expression_df)):\n",
    "        p_vals_list = []\n",
    "        phe = expression_df.iloc[i]\n",
    "        for index in genotypes_df.index:\n",
    "            gen = genotypes_df.loc[index]\n",
    "            non_others_indexes = gen[(((gen == 'B') | (gen == 'D')) | (gen == 'H'))].index\n",
    "            gen, phe = gen[non_others_indexes], phe[non_others_indexes].astype(float)\n",
    "            gen = gen.map({'B': 2, 'H': 1, 'D': 0}).astype(int)\n",
    "            p_vals_list.append(lrs(gen, phe)[3])\n",
    "        strains_dict[phe['data']] = p_vals_list\n",
    "    raw_mat = pd.DataFrame(strains_dict)\n",
    "    raw_mat = raw_mat.T\n",
    "    raw_mat.index = expression_df['data']\n",
    "    raw_mat.columns = genotypes_df.index\n",
    "    stacked = raw_mat.stack()\n",
    "    stacked = pd.Series(fdrcorrection(stacked, 0.05)[1], index=stacked.index)\n",
    "    final_mat = stacked.unstack()\n",
    "    final_mat.reset_index(inplace=True)\n",
    "    final_mat['min_p_val'] = final_mat.min(axis=1)\n",
    "    final_mat = final_mat[final_mat['min_p_val'] < alpha]\n",
    "    final_mat.drop('min_p_val', axis='columns', inplace=True)\n",
    "    final_mat.set_index('data', inplace=True)\n",
    "    return final_mat\n",
    "\n",
    "\n",
    "def create_gene_map_loc(mgi_df, p_vals_df):\n",
    "    mgi_df = deepcopy(mgi_df)\n",
    "    mgi_df.index = mgi_df['marker symbol']\n",
    "    mgi_df = mgi_df[['representative genome chromosome', 'representative genome start', 'representative genome end']]\n",
    "    relevant_gene_loc = mgi_df[mgi_df.index.isin(p_vals_df.index)]\n",
    "    relevant_gene_loc.columns = ['chromosome', 'start', 'end']\n",
    "    mgi_df = mgi_df.dropna(inplace=False)\n",
    "    return relevant_gene_loc, mgi_df\n",
    "\n",
    "\n",
    "def cis_trans(snp, gene):\n",
    "    if int(gene['chromosome']) == snp['Chr_Build37']:\n",
    "        position_snp = snp['Build37_position']\n",
    "        if np.abs(position_snp - gene['start']) <= 2000000 or np.abs(position_snp - gene['end']) <= 2000000:\n",
    "            return 'cis'\n",
    "        else:\n",
    "            return 'trans'\n",
    "    return 'trans'\n",
    "\n",
    "\n",
    "def all_loci(gene, genotypes_df):\n",
    "    d = {}\n",
    "    for i in range(len(genotypes_df)):\n",
    "        d[i] = cis_trans(genotypes_df.iloc[i], gene)\n",
    "    return pd.Series(d)\n",
    "\n",
    "\n",
    "def cis_trans_for_all_loci(genes, genotypes_df, every_loci):\n",
    "    d = {}\n",
    "    for i in range(len(genes)):\n",
    "        d[genes.iloc[i].name] = (all_loci(genes.iloc[i], genotypes_df))\n",
    "    d = pd.DataFrame(d).T\n",
    "    d.index.rename('marker symbol', inplace=True)\n",
    "    d = d.rename({i: every_loci[i] for i in range(len(list(d.columns)))}, axis='columns')\n",
    "    return d\n",
    "\n",
    "\n",
    "def count_significant_cis_trans(p_values, cis_trans_s):\n",
    "    combined = pd.DataFrame(p_values.unstack().rename('p_values')).join(cis_trans_s.rename_axis('data').unstack().rename('cis-trans'))\n",
    "    counts_df = pd.DataFrame(combined.loc[combined.p_values <= 0.05, 'cis-trans']).reset_index()\n",
    "    counts_df.rename({'level_0': 'snp'}, axis=1, inplace=True)\n",
    "    val_counts = counts_df['cis-trans'].value_counts()\n",
    "    counts_dict = {'Significant': val_counts['trans'] + val_counts['cis'], 'Cis': val_counts['cis'], 'Trans': val_counts['trans']}\n",
    "    return counts_df, counts_dict\n",
    "\n",
    "\n",
    "def eQTL_association_df(p_values, filtered_genotypes_df):\n",
    "    p_values2 = p_values.reset_index()\n",
    "    p_values2 = p_values2.drop('data', axis=1)\n",
    "    sum_of_eQTL = (p_values2 < 0.05).sum()\n",
    "    eQTL_df = pd.DataFrame({'Count': sum_of_eQTL})\n",
    "    filtered_genotypes_df['Chr_Build37'].index = eQTL_df.index\n",
    "    return pd.DataFrame(eQTL_df).join(filtered_genotypes_df['Chr_Build37'])\n",
    "\n",
    "\n",
    "def plot_eQTL(eQTL_df):\n",
    "    eQTL_df['place'] = range(len(eQTL_df))\n",
    "    eQTL_sorted = eQTL_df.sort_values('Chr_Build37')\n",
    "    eQTL_sorted.groupby('Chr_Build37')\n",
    "    eQTL_df.reset_index(inplace=True, drop=True)\n",
    "    eQTL_df['index'] = eQTL_df.index\n",
    "    sns.set(font_scale=1.25)\n",
    "    sns.set_style('dark')\n",
    "    plot = sns.relplot(data=eQTL_df, x='index', y='Count', aspect=3.7, hue='Chr_Build37', palette='dark', legend=None, s=200)\n",
    "    chrom_df = eQTL_df.groupby('Chr_Build37')['index'].median()\n",
    "    plot.ax.set_xlabel('Chromosome No.')\n",
    "    plot.ax.set_xticks(chrom_df)\n",
    "    plot.ax.set_xticklabels(chrom_df.index)\n",
    "    plot.fig.suptitle('Number of genes associated with each eQTL')\n",
    "    plot.set(yticks=[i for i in range(0,4)])\n",
    "    plt.savefig('genes-eQTLS association.png')\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def create_eQTL_ditribution_data(p_values, cis_trans):\n",
    "    df = pd.DataFrame(p_values.unstack().rename(\"p_values\")).join(\n",
    "        cis_trans.rename_axis(\"data\").unstack().rename(\"cis_trans\"))\n",
    "    filter_arr = lambda x: df.loc[df['cis_trans'] == x, 'p_values'].to_numpy()\n",
    "    cis_arr, trans_arr = filter_arr('cis'), filter_arr('trans')\n",
    "    log_cis_arr, log_trans_arr = -np.log10(cis_arr), -np.log10(trans_arr)\n",
    "    significant_cis_arr, significant_trans_arr = cis_arr[cis_arr <= 0.05], trans_arr[trans_arr <= 0.05]\n",
    "    log_significant_cis_arr, log_significant_trans_arr = -np.log10(significant_cis_arr), -np.log10(\n",
    "        significant_trans_arr)\n",
    "    return log_significant_cis_arr, log_significant_trans_arr, log_cis_arr, log_trans_arr\n",
    "\n",
    "\n",
    "def plot_distribution_graph(cis, trans):\n",
    "    cis, trans = np.array(list(cis)), np.array(list(trans))\n",
    "    print(cis)\n",
    "    print(trans)\n",
    "    g = sns.kdeplot(cis, color=\"r\", label='Cis')\n",
    "    sns.kdeplot(trans, color=\"b\", label='Trans', bw=0.3)\n",
    "    plt.legend(loc='upper right')\n",
    "    g.set_ylabel('Density')\n",
    "    g.set_xlabel('-log p value')\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def create_snp_gene_df(sig_df, genotypes_df, relevant_gene_map_loc_df):\n",
    "    df = relevant_gene_map_loc_df.merge(sig_df, right_on='data', left_index=True)\n",
    "    df['gene location'] = (df['start'] + df['end']) / 2\n",
    "    df.drop(['start', 'end'], axis=1, inplace=True)\n",
    "    df['snp'] = df['snp'].astype(int)\n",
    "    df = df.merge(genotypes_df[['Chr_Build37', 'Build37_position']], left_on='snp', right_index=True)\n",
    "    columns = ['gene name', 'gene location', 'gene chromosome', 'snp name', 'snp location', 'snp chromosome', 'cis-trans']\n",
    "    df.rename({'data': 'gene name', 'chromosome': 'gene chromosome', 'Chr_Build37': 'snp chromosome', 'Build37_position': 'snp location', 'snp': 'snp name'}, axis=1,\n",
    "               inplace=True)\n",
    "    df = df[columns]\n",
    "    return df\n",
    "\n",
    "\n",
    "def find_max_pos(chr_data):\n",
    "    ser = chr_data.groupby('representative genome chromosome')['representative genome end'].max()\n",
    "    for i in range(2, len(ser) + 1):\n",
    "        ser[i] += ser[i - 1]\n",
    "    return ser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e2e623cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_file = pd.read_excel('genotypes.xls', skiprows=1).drop(0, axis=1)\n",
    "txt_file = \"dendritic_PAM_stimulation.txt\"\n",
    "data = pd.read_csv(txt_file, sep='\\t', header=0)\n",
    "strains = mean_data(data)\n",
    "genotypes = filtering(genotypes_file, strains.columns[1:])\n",
    "\n",
    "# p_vals = association_model(genotypes, strains)  ## Takes a long time, csv attached and imported in the line below:\n",
    "p_vals = pd.read_csv(\"after association test.csv\", index_col=[0])\n",
    "\n",
    "mgi_file = \"MGI_Coordinates.Build37.rpt.txt\"\n",
    "mgi = pd.read_csv(mgi_file, sep='\\t', header=0)\n",
    "loci = p_vals.columns\n",
    "\n",
    "relevant, loc2 = create_gene_map_loc(mgi, p_vals)\n",
    "\n",
    "cis_trans_df = cis_trans_for_all_loci(relevant, genotypes, loci)\n",
    "\n",
    "### Question 1 ###\n",
    "counts = count_significant_cis_trans(p_vals, cis_trans_df)\n",
    "sig = counts[0]\n",
    "counts_val = counts[1]\n",
    "\n",
    "### Question 3 ###\n",
    "sig_cis, sig_trans, cis, trans = create_eQTL_ditribution_data(p_vals, cis_trans_df)\n",
    "\n",
    "### Question 4 ###\n",
    "\n",
    "TMP = create_snp_gene_df(sig, genotypes, relevant)\n",
    "chr_data = loc2[(loc2['representative genome chromosome'] != 'Y') & (loc2['representative genome chromosome'] != 'MT')]\n",
    "chr_data['representative genome chromosome'] = chr_data['representative genome chromosome'].astype(float)\n",
    "max_pos = find_max_pos(chr_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6b945612",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(range(1, 21))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b8b0663",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cis_trans_gene_position(chr_df, max_pos_ser):\n",
    "    chr_no = list(range(1,21))\n",
    "    fig, ax = plt.subplots()\n",
    "    cis = chr_df.loc[chr_df['cis-trans'] == 'cis']\n",
    "    x = cis['normalized location of snp']\n",
    "    y = cis['normalized location of gene']\n",
    "    cis = ax.plot(x, y, 'o', label='cis', markersize=0)\n",
    "    trans = chr_df.loc[chr_df['cis-trans'] == 'trans']\n",
    "    x = trans['normalized location of snp']\n",
    "    y = trans['normalized location of gene']\n",
    "    trans = ax.plot(x, y, 'o', label='trans', markersize=0)\n",
    "    ax.set_ylabel('Gene Location')\n",
    "    ax.set_xlabel('SNP Location')\n",
    "    ax.legend()\n",
    "    ax.tick_params(axis='x', labelsize=9)\n",
    "    ax.tick_params(axis='y', labelsize=9)\n",
    "    #### MAYBE ADD CONDITION DUNNO NEED TO FIGURE EFFECT:\n",
    "    ax.set_xticks(max_pos_ser)\n",
    "    plt.xticks(max_pos_ser, chr_no)\n",
    "    ax.set_xticks(max_pos_ser)\n",
    "    plt.yticks(max_pos_ser, chr_no)\n",
    "    \n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
